README.md
# Chattin

Chattin is an application to talk with other people. You can join different rooms to talk with different people.

![Logo](https://github.com/RobinSchapendonk/Chattin/blob/master/build/icon.png)

## Register

Open the app, enter an username and a password, then click "register" to register

## Rooms

Chattin uses rooms to communicate with eachother, you both need to be in the same room to see the messages from eachother.

## About the project

This is an [electron](https://www.electronjs.org/) application, which uses e2e encryption to chat with other people.

CHANGELOG.MD
# v0.0.5
More logging
Server change

# v0.0.4
No email required

# v0.0.3
End-to-end (e2e) encryption

# v0.0.2
Rewrite/fixed a lot of code, jwt login

# v0.0.1
Inital commit

index.js
const electron = require('electron');
const path = require('path');
const log = require('electron-log');
const { autoUpdater } = require('electron-updater');

const { app, BrowserWindow, Menu, ipcMain } = electron;
const SaveData = require('./js/SaveData.js');
const currentData = new SaveData({
	configName: 'current',
});

let server = app.isPackaged ? 'http://142.54.191.92:1846' : 'http://localhost:3000';
currentData.set('server', server);

autoUpdater.logger = log;
autoUpdater.logger.transports.file.level = 'info';
log.info('App starting...');

let mainWindow;
let loginWindow;
let currentWindow;

const gotTheLock = app.requestSingleInstanceLock();

if (!gotTheLock && app.isPackaged) {
	app.quit();
} else {
	app.on('second-instance', () => {
		if (currentWindow) {
			if (currentWindow.isMinimized()) currentWindow.restore();
			currentWindow.focus();
		}
	});

	app.on('ready', () => {
		loginWindow = new BrowserWindow({ 
			minWidth: 400, 
			minHeight: 400, 
			icon: path.join(__dirname, 'build/icon.png'), 
			webPreferences: {
				nodeIntegration: true,
				contextIsolation: false
			}
		});
		loginWindow.loadURL(`file://${path.join(__dirname, 'html/login.html')}`);
		loginWindow.on('closed', () => {
			loginWindow = null;
		});
		const loginMenu = Menu.buildFromTemplate(loginMenuTemplate);
		Menu.setApplicationMenu(loginMenu);
		currentWindow = loginWindow;

		if(app.isPackaged) {
			log.info('starting update check');
			autoUpdater.checkForUpdates();
		} else {
			currentWindow.webContents.openDevTools();
		}

	});


	// Menu options bar
	const mainMenuTemplate = [];
	const loginMenuTemplate = [];

	ipcMain.on('open:index', (event, data) => {
		mainWindow = new BrowserWindow({ 
			show: false, 
			minWidth: 400, 
			minHeight: 400, 
			icon: path.join(__dirname, 'build/icon.png'), 
			webPreferences: { 
				nodeIntegration: true,
				contextIsolation: false,
			}
		});
		if(!app.isPackaged) mainWindow.webContents.openDevTools();
		mainWindow.maximize();
		mainWindow.show();
		mainWindow.loadURL(`file://${path.join(__dirname, 'html/index.html')}`);
		mainWindow.on('closed', () => {
			app.quit();
		});
		const mainMenu = Menu.buildFromTemplate(mainMenuTemplate);
		Menu.setApplicationMenu(mainMenu);
		currentWindow = mainWindow;

		mainWindow.webContents.once('dom-ready', () => {
			mainWindow.show();
			mainWindow.webContents.send('start', data);
		});

		if (loginWindow) loginWindow.close();
	});

	ipcMain.on('close', () => {
		currentWindow.close();
	});

	ipcMain.on('focus', () => {
		if (currentWindow.isMinimized()) currentWindow.restore();
		currentWindow.focus();
	});

	ipcMain.on('getPath', (event, data) => {
		const userDataPath = electron.app.getPath(data);
		currentWindow.webContents.send('getPath', userDataPath);
	});

	// If mac, add empty object to menu
	if(process.platform == 'darwin') {
		mainMenuTemplate.unshift({});
	}

	// -------------------------------------------------------------------
	// Auto updates
	// -------------------------------------------------------------------
	autoUpdater.on('checking-for-update', () => {
	// if (currentWindow) currentWindow.webContents.send('alert', 'Checking for updates');
	});
	autoUpdater.on('update-available', (ev, info) => {
		if (currentWindow) currentWindow.webContents.send('alert', 'Download new update!');
		if (info) log.info('update-available info', info);
	});
	autoUpdater.on('update-not-available', (ev, info) => {
	// if (currentWindow) currentWindow.webContents.send('alert', 'Update not available');
		if (info) log.info('update-not-available info', info);
	});
	autoUpdater.on('error', (ev, err) => {
		if (currentWindow) currentWindow.webContents.send('alert', 'Error in auto updater!');
		if (err) log.info('update-error error', err);
	});
	autoUpdater.on('update-downloaded', (ev, info) => {
		if (currentWindow) currentWindow.webContents.send('alert', 'Restarting in 3 seconds for update!');
		if (info) log.info('update-download info', info);
		setTimeout(() => {
			autoUpdater.quitAndInstall();
		}, 3000);
	});
}

js/main.js
$(function() {
	const { ipcRenderer } = require('electron');
	const { sessionKeys, loadEncryptionObjects, generateSessionKeys, pack, unpack } = require('../js/crypt.js');
	const SaveData = require('../js/SaveData.js');
	const currentData = new SaveData({
		configName: 'current',
	});

	let token, uid, room, server;

	ipcRenderer.on('start', (event, data) => {
		token = data.token;
		uid = data.uid;
		room = data.room;
	});

	const getPrevious = setInterval(() => {
		if (!server) {
			server = currentData.get('server');
		} else {
			generateSessionKeys();
			ipcRenderer.send('log', 'Generated key');
			
			document.getElementById('chat-errors').innerHTML = 'Connecting to the server';
			let socket = io(server);

			socket.on('authenticated', () => {
				setTimeout(() => {
					socket.emit('joinRoom', room);
					socket.emit('init-session', {
						status: true,
						key: sessionKeys.client.public
					});
				}, 1000);
			});

			socket.on('connect', () => {
				error(false, 'Connected to the server');
			});

			socket.on('disconnect', () => {
				error(true, 'Disconnected from server, please restart Chattin!');
			});
			socket.on('connect_failed', () => {
				error(true, 'Disconnected from server, please restart Chattin!');
			});
			socket.on('connect_error', () => {
				error(true, 'Disconnected from server, please restart Chattin!');
			});
			socket.on('slowmode', (data, timeout) => {
				const decrypted = JSON.parse(unpack(data));
				$('#send_message').css('border-color', 'red');
				$('#send_message').val(decrypted.message);
				setTimeout(() => {
					$('#send_message').css('border-color', 'black');
				}, timeout);
			});
			socket.on('init-session', (data) => {
				loadEncryptionObjects(data.key);
			});
			$('#send_message_form').submit((e) => {
				e.preventDefault();
				const msg = $('#send_message').val();
				if (!msg) return;

				const data = pack({
					message: msg,
					uid,
					room,
				});
				socket.emit('message', data);

				$('#send_message').val('');
				return false;
			});
			$('#change_room_form').submit((e) => {
				e.preventDefault();
				const newRoom = $('#change_room').val();
				if (!newRoom) return;
				socket.emit('joinRoom', newRoom);
				room = newRoom;
				$('#change_room').val('');
				$('#chat-messages').text('');
				return false;
			});
			socket.on('message', (data) => {
				const decrypted = JSON.parse(unpack(data));
				$('#chat-messages').append($('<p style="white-space: pre-line">').text(`${decrypted.username}:\n${decrypted.message}`));
				$('#chat').animate({ scrollTop: $('#chat').prop('scrollHeight') });
				if (data.uid !== uid) {
					const audio = new Audio('../sounds/message.mp3');
					audio.play();
					const myNotification = new Notification(decrypted.username, {
						body: decrypted.message,
					});
					myNotification.onclick = () => {
						ipcRenderer.send('focus');
					};
				}
			});
			socket.on('error', (msg) => {
				ipcRenderer.send(msg);
			});
			socket.on('users', (users) => {
				$('#users').html(`${users} user${users > 1 ? 's' : ''} in chat`);
			});
			socket.on('details', (data) => {
				$('#server').html(`Room: ${data.room}<br>Server: ${data.server}`);
			});

			socket.emit('authenticate', { token });
			token = null;

			clearInterval(getPrevious);
		}
	}, 500)
});

const error = (errorBool, msg) => {
	if(errorBool) {
		ipcRenderer.send('log', msg);
		$('#chat-errors').text(msg);
		$('#send_message').prop('placeholder', msg);
		$('#send_message').prop('disabled', true);
	} else {
		ipcRenderer.send('log', msg);
		$('#chat-errors').text('');
		$('#send_message').prop('placeholder', 'Send a message.');
		$('#send_message').prop('disabled', false);
	}
}

js/crypt.js
const md5 = require('./lib/md5.min.js');
const aesjs = require('aes-js');

let SECURITY_LEVEL = 4096;

let sessionKeys = {
	client: {},
	server: {},
};

let encrypter, decrypter;

function generateRandomString(length) {
	let text = '';
	const charset = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
	for (let i = 0; i < length; i++) {
		text += charset.charAt(Math.floor(Math.random() * charset.length));
	}
	return text;
}

function generateSessionKeys() {
	const crypt = new JSEncrypt({ default_key_size: SECURITY_LEVEL });
	crypt.getKey();

	sessionKeys.client = {
		'private': crypt.getPrivateKey(),
		'public': crypt.getPublicKey(),
	};
}

function loadEncryptionObjects(serverPublicKey) {
	sessionKeys.server.public = serverPublicKey;

	encrypter = new JSEncrypt();
	encrypter.setKey(sessionKeys.server.public);

	decrypter = new JSEncrypt();
	decrypter.setKey(sessionKeys.client.private);
}

const aes = {
	encrypt: function(secret, text) {
		const secretHash = md5(secret);
		const key = aesjs.utils.utf8.toBytes(secretHash);
		const textBytes = aesjs.utils.utf8.toBytes(text);
		const aesCtr = new aesjs.ModeOfOperation.ctr(key);
		const encryptedBytes = aesCtr.encrypt(textBytes);
		return encryptedBytes;
	},

	decrypt: function(secret, encryptedBytes) {
		const encryptedData = [];
		const keys = Object.keys(encryptedBytes);
		for (let i = 0; i < keys.length; i++) {
			encryptedData[i] = encryptedBytes[keys[i]];
		}
		const secretHash = md5(secret);
		const key = aesjs.utils.utf8.toBytes(secretHash);
		const aesCtr = new aesjs.ModeOfOperation.ctr(key);
		const decryptedBytes = aesCtr.decrypt(encryptedData);
		return aesjs.utils.utf8.fromBytes(decryptedBytes);
	},

	generateKey: function() {
		return generateRandomString(32);
	},
};

function pack(data) {
	const packedData = {};
	const aesKey = aes.generateKey();
	try {
		packedData.key = encrypter.encrypt(aesKey);
		packedData.encrypted = aes.encrypt(aesKey, JSON.stringify(data));
		return packedData;
	} catch (dataEncryptionException) {
		console.log('failed to pack message: ' + dataEncryptionException.message);
		return {};
	}
}

function unpack(data) {
	const aesKey = decrypter.decrypt(data.key);
	return aes.decrypt(aesKey, data.encrypted);
}

module.exports = {
	sessionKeys,
	loadEncryptionObjects,
	generateSessionKeys,
	pack,
	unpack,
}

js/login.js
const electron = require('electron');
const { ipcRenderer } = electron;
const SaveData = require('../js/SaveData.js');
const savedata = new SaveData({
	configName: 'user-preferences',
});
const currentData = new SaveData({
	configName: 'current',
});

let server = undefined;
let oldUsername = undefined;
let oldRoom = undefined;

const getPrevious = setInterval(() => {
	if (!oldUsername) {
		oldUsername = savedata.get('username');
		if (oldUsername) document.getElementById('Lusername').value = oldUsername;
	
		oldRoom = savedata.get('room');
		if (oldRoom) document.getElementById('Lroom').value = oldRoom;

		server = currentData.get('server');
		if (!server) document.getElementById('response').innerHTML = 'There are problems connecting to the server!';
	} else {
		clearInterval(getPrevious);
	}
}, 500)

const login = async () => {
	const username = document.getElementById('Lusername').value;
	const password = document.getElementById('Lpassword').value;
	let room = document.getElementById('Lroom').value;
	if (!room) room = 'public';
	
	const valid = validate(username, password, room);
	if (!valid.success) return alert(valid.message);

	const body = {
		username,
		password,
		room,
	};
	fetch(`${server}/login`, {
		method: 'POST',
		body: JSON.stringify(body),
		headers: { 'Content-Type': 'application/json' },
	}).then(res => res.json())
		.then(json => {
			if (!json.success) {
				return document.getElementById('response').innerHTML = json.message;
			} else {
				savedata.set('username', username);
				savedata.set('room', room);
				ipcRenderer.send('open:index', { uid: json.uid, room, token: json.token, uid: json.uid });
			}
		}).catch(() => {
			document.getElementById('response').innerHTML = 'There are problems connecting to the server!';
		});
}
const register = async () => {
	const username = document.getElementById('Rusername').value;
	const password = document.getElementById('Rpassword').value;
	let room = document.getElementById('Rroom').value;
	if (!room) room = 'public';

	const valid = validate(username, password, room);
	if (!valid.success) return alert(valid.message);

	const body = {
		username,
		password,
		room,
	};
	fetch(`${server}/register`, {
		method: 'POST',
		body: JSON.stringify(body),
		headers: { 'Content-Type': 'application/json' },
	}).then(res => res.json())
		.then(json => {
			if (!json.success) {return document.getElementById('response').innerHTML = json.message;} else {
				savedata.set('username', username);
				savedata.set('room', room);
				ipcRenderer.send('open:index', { uid: json.uid, room, token: json.token, uid: json.uid });
			}
		}).catch(() => {
			document.getElementById('response').innerHTML = 'There are problems connecting to the server!';
		});
}
const validate = (username, password, room) => {
	if (!username) {
		return { success: false, message: 'Missing your username!' };
	} else if (username.length < 1 || username.length > 32) {
		return { success: false, message: 'Your username can\'t be longer than 32 characters' };
	} else if (!password) {
		return { success: false, message: 'Missing your password!' };
	} else if (!room) {
		return { success: false, message: 'Missing the room!' };
	} else if (room.length < 0 || room.length > 32) {
		return { success: false, message: 'Your room name can\'t be longer than 32 characters!' };
	} else {
		return { success: true };
	}
}

const change = (block) => {
	const other = block == 'login' ? 'register' : 'login';
	document.getElementById(block).style = 'display: block';
	document.getElementById(other).style = 'display: none';
}

const error = (errorBool, msg) => {
	if (errorBool) {
		document.getElementById('response').innerHTML = msg;
	} else {
		document.getElementById('response').innerHTML = '';
	}
}